src/restconfig/src/test/java/org/geoserver/rest/catalog/FeatureTypeControllerTest.java

```diff
@@ -943,4 +945,110 @@ public void testPostWithAttributes() throws Exception {
         FeatureTypeInfo ft = getCatalog().getFeatureTypeByName("sf", typeName2);
         assertNotNull(ft.getFeatureType());
     }
+
+    @Test
+    public void testPostWithRestrictedAttributes() throws Exception {
+        String xml = getAsString(BASEPATH + "/workspaces/sf/featuretypes/PrimitiveGeoFeature.xml");
+
+        // create a new sibling feature type with restricted attributes
+        String typeWitRestrictedAttributesName = "PrimitiveGeoFeatureRestricted";
+        String xmlWithRestrictedAttributes = xml.replace(
+                        "<name>PrimitiveGeoFeature</name>", "<name>" + typeWitRestrictedAttributesName + "</name>")
+                .replace(
+                        "<name>intProperty</name>",
+                        """
+                    <name>intProperty</name>
+                    <options>
+                        <int>1</int>
+                        <int>2</int>
+                        <int>3</int>
+                    </options>""")
+                .replace(
+                        "<name>decimalProperty</name>",
+                        """
+                    <name>decimalProperty</name>
+                    <range>
+                        <min>3.14</min>
+                        <max>99.99</max>
+                    </range>""");
+
+        MockHttpServletResponse sr =
+                postAsServletResponse(BASEPATH + "/workspaces/sf/featuretypes", xmlWithRestrictedAttributes);
+        assertEquals(201, sr.getStatus());
+
+        Document dom = getAsDOM(BASEPATH + "/workspaces/sf/featuretypes/" + typeWitRestrictedAttributesName + ".xml");
+
+        /* options restriction */
+        NodeList optionsNodes = xp.getMatchingNodes("//attribute[name='intProperty']/options/int", dom);
+        List<String> options = IntStream.range(0, optionsNodes.getLength())
+                .mapToObj(i -> optionsNodes.item(i).getTextContent().trim())
+                .toList();
+        assertEquals(List.of("1", "2", "3"), options);
+
+        /* range restriction */
+        assertXpathEvaluatesTo("3.14", "//attribute[name='decimalProperty']/range/min", dom);
+        assertXpathEvaluatesTo("99.99", "//attribute[name='decimalProperty']/range/max", dom);
+    }
+
+    @Test
+    public void testPutWithRestrictedAttributes() throws Exception {
+        String xml = getAsString(BASEPATH + "/workspaces/sf/featuretypes/PrimitiveGeoFeature.xml");
+
+        // update feature type with restricted attributes
+        String xmlWithRestrictedAttributes = xml.replace(
+                        "<name>intProperty</name>",
+                        """
```

Comment:
Just an observation, when you backport mind that code will have to be java 11 compatible and text block will have to be converted to string concatenation.